core/src/main/scala/tech/beshu/ror/es/DataStreamService.scala

```diff
@@ -16,42 +16,122 @@
+import tech.beshu.ror.utils.ScalaOps.retryBackoffEither
+
+import scala.concurrent.duration.*
 
 trait DataStreamService {
 
-  final def fullySetupDataStream(settings: DataStreamSettings): Task[Unit] = {
+  final def fullySetupDataStream(settings: DataStreamSettings): Task[DataStreamSetupResult] = {
     for {
-      _ <- createIndexLifecyclePolicy(settings.lifecyclePolicy)
-      _ <- createComponentTemplateForMappings(settings.mappings)
-      _ <- createComponentTemplateForIndex(settings.componentSettings)
-      _ <- createIndexTemplate(settings.templateSettings)
-      _ <- createDataStream(settings.dataStreamName)
-    } yield ()
-  }
+      _ <- createIfAbsent(
+        checkIfResourceExists = checkIndexLifecyclePolicyExists(settings.lifecyclePolicy.id),
+        createResource = createIndexLifecyclePolicy(settings.lifecyclePolicy),
+        onNotAcknowledged = DataStreamSetupResult.Failure(s"Unable to determine if the index lifecycle policy with ID '${settings.lifecyclePolicy.id.show}' has been created")
+      )
+      _ <- createIfAbsent(
+        checkIfResourceExists = checkComponentTemplateExists(settings.mappings.templateName),
+        createResource = createComponentTemplateForMappings(settings.mappings),
+        onNotAcknowledged = DataStreamSetupResult.Failure(s"Unable to determine if component template with ID '${settings.mappings.templateName.show}' has been created")
+      )
+      _ <- createIfAbsent(
+        checkIfResourceExists = checkComponentTemplateExists(settings.componentSettings.templateName),
+        createResource = createComponentTemplateForIndex(settings.componentSettings),
+        onNotAcknowledged = DataStreamSetupResult.Failure(s"Unable to determine if component template with ID '${settings.componentSettings.templateName.show}' has been created")
+      )
+      _ <- createIfAbsent(
+        checkIfResourceExists = checkIndexTemplateExists(settings.templateSettings.templateName),
+        createResource = createIndexTemplate(settings.templateSettings),
+        onNotAcknowledged = DataStreamSetupResult.Failure(s"Unable to determine if index template with ID '${settings.templateSettings.templateName.show}' has been created")
+      )
+      _ <- createIfAbsent(
+        checkIfResourceExists = checkDataStreamExists(settings.dataStreamName),
+        createResource = createDataStream(settings.dataStreamName),
+        onNotAcknowledged = DataStreamSetupResult.Failure(s"Unable to determine if data stream with ID '${settings.dataStreamName.show}' has been created")
+      )
+    } yield DataStreamSetupResult.Success
+  }.merge
 
   def checkDataStreamExists(dataStreamName: DataStreamName.Full): Task[Boolean]
 
   protected def createDataStream(dataStreamName: DataStreamName.Full): Task[CreationResult]
 
+  protected def checkIndexLifecyclePolicyExists(policyId: NonEmptyString): Task[Boolean] = ??? // default implementation to be removed after porting changes to all modules
```

Comment:
I'm leaving a comment as a reminder about this, just in case. Tests would probably catch it anyway, but it's better to be safe than sorry ;)